File: killrvideo-commons/src/main/java/com/killrvideo/grpc/AbstractSingleServiceGrpcServer.java
package com.killrvideo.grpc;

import java.util.Optional;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.killrvideo.conf.KillrVideoConfiguration;
import com.killrvideo.discovery.ServiceDiscoveryDaoEtcd;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.ServerServiceDefinition;

/**
 * Support class to build GRPC Server per service.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
public abstract class AbstractSingleServiceGrpcServer {
    
    /** Some logger. */
    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractSingleServiceGrpcServer.class);
   
    /** Global Configuration. s*/
    @Inject
    protected KillrVideoConfiguration killrVideoConfig;
    
    /** Connectivity to ETCD Service discovery. */
    @Inject
    protected ServiceDiscoveryDaoEtcd serviceDiscoveryDao;
    
    /** GRPC Server to start. */
    protected Server grpcServer;
    
    /** Port to be allocated dynamically based on ETCD. */
    protected static int grpcServerPort = 0;
    
    /** Service Name. */
    protected abstract String getServiceName();
    
    /** Lock target port. */
    protected abstract int getDefaultPort();
    
    /** Service Definition. */
    protected abstract ServerServiceDefinition getService();
   
    /**
     * Start
     */
    @PostConstruct
    public void startGrpcServer() throws Exception {
        LOGGER.info("Initializing Comment Service");
        grpcServerPort = getDefaultPort();
        Optional<Integer> maxUsedPort = serviceDiscoveryDao.lookupServicePorts(getServiceName(), 
                killrVideoConfig.getApplicationHost());
        maxUsedPort.ifPresent(integer -> grpcServerPort = integer + 1);
        grpcServer = ServerBuilder.forPort(grpcServerPort)
                              .addService(getService())
                              .build();
        Runtime.getRuntime().addShutdownHook(new Thread(this::stopGrpcServer));
        grpcServer.start();
        LOGGER.info("[OK] Grpc Server started on port: '{}'", grpcServerPort);
        serviceDiscoveryDao.register(getServiceName(), 
                killrVideoConfig.getApplicationHost(), grpcServerPort);
    }
    
    @PreDestroy
    public void stopGrpcServer() {
        LOGGER.info("Calling shutdown for GrpcServer");
        serviceDiscoveryDao.register(getServiceName(), 
                killrVideoConfig.getApplicationHost(), grpcServerPort);
        grpcServer.shutdown();
    }
    
}

File: killrvideo-commons/src/main/kotlin/com/killrvideo/grpc/AbstractSingleServiceGrpcServer.kt
package com.killrvideo.grpc

import com.killrvideo.conf.KillrVideoConfiguration
import com.killrvideo.discovery.ServiceDiscoveryDaoEtcd
import com.killrvideo.grpc.AbstractSingleServiceGrpcServer
import io.grpc.Server
import io.grpc.ServerBuilder
import io.grpc.ServerServiceDefinition
import org.slf4j.LoggerFactory
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy
import javax.inject.Inject

/**
 * Support class to build GRPC Server per service.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
abstract class AbstractSingleServiceGrpcServer {
    /** Global Configuration. s */
    @Inject
    protected var killrVideoConfig: KillrVideoConfiguration? = null

    /** Connectivity to ETCD Service discovery.  */
    @Inject
    protected var serviceDiscoveryDao: ServiceDiscoveryDaoEtcd? = null

    /** GRPC Server to start.  */
    protected var grpcServer: Server? = null

    /** Service Name.  */
    protected abstract val serviceName: String?

    /** Lock target port.  */
    protected abstract val defaultPort: Int

    /** Service Definition.  */
    protected abstract val service: ServerServiceDefinition?

    /**
     * Start
     */
    @PostConstruct
    @Throws(Exception::class)
    fun startGrpcServer() {
        LOGGER.info("Initializing Comment Service")
        grpcServerPort = defaultPort
        val maxUsedPort = serviceDiscoveryDao!!.lookupServicePorts(
            serviceName!!,
            killrVideoConfig!!.applicationHost
        )
        maxUsedPort.ifPresent { integer: Int -> grpcServerPort = integer + 1 }
        grpcServer = ServerBuilder.forPort(grpcServerPort)
            .addService(service)
            .build()
        Runtime.getRuntime().addShutdownHook(Thread { stopGrpcServer() })
        grpcServer!!.start()
        LOGGER.info("[OK] Grpc Server started on port: '{}'", grpcServerPort)
        serviceDiscoveryDao!!.register(
            serviceName!!,
            killrVideoConfig!!.applicationHost, grpcServerPort
        )
    }

    @PreDestroy
    fun stopGrpcServer() {
        LOGGER.info("Calling shutdown for GrpcServer")
        serviceDiscoveryDao!!.register(
            serviceName!!,
            killrVideoConfig!!.applicationHost, grpcServerPort
        )
        grpcServer!!.shutdown()
    }

    companion object {
        /** Some logger.  */
        private val LOGGER = LoggerFactory.getLogger(AbstractSingleServiceGrpcServer::class.java)

        /** Port to be allocated dynamically based on ETCD.  */
        protected var grpcServerPort = 0
    }
}
